src/config/mod.rs

```diff
@@ -53,53 +51,211 @@ impl MessageHubOptions {
     /// configuration.
     pub async fn get() -> Result<MessageHubOptions, AstarteMessageHubError> {
-        let base_config = read_options_from_base_locations()?;
-        if base_config.is_valid() {
-            return Ok(base_config);
-        }
+        let paths = [file::CONFIG_FILE_NAME, "/etc/message-hub/config.toml"]
+            .iter()
+            .map(Path::new)
+            .filter(|f| f.exists());
 
-        let store_directory = base_config.store_directory.clone();
-        let path = Path::new(&store_directory).join(CONFIG_FILE_NAME);
+        if let Some(path) = paths.into_iter().next() {
+            info!("Found configuration file {path:?}");
 
-        if path.exists() {
-            return file::read_options(&path);
-        }
+            let (msg_hub_opt, redirect_dir) = file::get_options(path)?;
+            if let Some(msg_hub_opt) = msg_hub_opt {
+                return Ok(msg_hub_opt);
+            }
+
+            let (tx, mut rx) = channel(1);
 
-        let (tx, mut rx) = channel(1);
+            let web_server =
+                HttpConfigProvider::new("127.0.0.1:40041", redirect_dir.clone(), tx.clone());
+            let protobuf_server =
+                ProtobufConfigProvider::new("[::1]:50051", redirect_dir.clone(), tx.clone()).await;
 
-        let web_server =
-            HttpConfigProvider::new("127.0.0.1:40041", store_directory.clone(), tx.clone());
-        let protobuf_server =
-            ProtobufConfigProvider::new("[::1]:50051", store_directory.clone(), tx.clone()).await;
+            rx.recv().await.unwrap();
 
-        rx.recv().await.unwrap();
+            web_server.stop().await;
+            protobuf_server.stop().await;
 
-        web_server.stop().await;
-        protobuf_server.stop().await;
-        file::read_options(&path)
+            let toml_str = std::fs::read_to_string(redirect_dir)?;
+            let err_msg = AstarteMessageHubError::FatalError("Invalid configuration.".to_string());
+            file::read_options_from_toml(&toml_str).ok_or(err_msg)
+        } else {
+            Err(AstarteMessageHubError::FatalError(
+                "Configuration file not found in base locations.".to_string(),
+            ))
+        }
     }
 
-    fn is_valid(&self) -> bool {
-        self.device_id.is_some()
-            && self.realm.is_some()
-            && self.pairing_url.is_some()
-            && (self.pairing_token.is_some() || self.credentials_secret.is_some())
+    pub fn is_valid(&self) -> bool {
+        let valid_cs = Some(true) == self.credentials_secret.clone().map(|e| !e.is_empty());
+        let valid_pt = Some(true) == self.pairing_token.clone().map(|e| !e.is_empty());
+        let valid_id = self.interfaces_directory.is_none()
+            || (Some(true)
+                == self.interfaces_directory.clone().map(|e| {
+                    let id_path = Path::new(&e);
+                    id_path.exists() && id_path.is_dir()
+                }));
```

Comment:
I found `valid_id` very confusing. I was thinking we were talking about the `device_id` here instead of the `interface_directory`. I think a clearer name is better than a short one for clarity's sake.
I suggest renaming them all.

```suggestion
        let valid_secret = Some(true) == self.credentials_secret.clone().map(|e| !e.is_empty());
        let valid_token = Some(true) == self.pairing_token.clone().map(|e| !e.is_empty());
        let valid_interface_dir = self.interfaces_directory.is_none()
            || (Some(true)
                == self.interfaces_directory.clone().map(|e| {
                    let id_path = Path::new(&e);
                    id_path.exists() && id_path.is_dir()
                }));
```

`valid_iface_dir` could be a shorter alternative to `valid_interface_dir` even though `iface` usually refers to network interfaces.